File: Week_01/python/1041_robot_bounded_in_circle.py
class Solution0:
    def isRobotBounded(self, instructions: str) -> bool:
        instructions *= 4
        x, y, direction = [0, 0, 0]

        for ch in instructions:
            if ch == 'L':
                direction = (direction - 1) % 4
            elif ch == 'R':
                direction = (direction + 1) % 4
            else:
                if direction == 0:
                    y += 1
                elif direction == 1:
                    x += 1
                elif direction == 2:
                    y -= 1
                else:
                    x -= 1

        return x == 0 and y == 0


class Solution1:
    def isRobotBounded(self, instructions: str) -> bool:
        instructions *= 4
        x, y, d = [0, 0, 0]
        DIRS = [[1,0], [0,1], [-1,0], [0, -1]]

        for ch in instructions:
            if ch == 'L':
                d = (d - 1) % 4
            elif ch == 'R':
                d = (d + 1) % 4
            else:
                x += DIRS[d][0]
                y += DIRS[d][1]

        return x == 0 and y == 0

File: Week_01/java/1041_robot_bounded_in_circle.java
class Solution{
    public boolean isRobotBounded(String instructions){
        int x = 0, y = 0, d = 0;
        int[][] directions= new int[][]{
                new int[]{1, 0}, new int[]{0, 1}, new int[]{-1, 0}, new int[]{0, -1}
        };

        for(int time=0; time < 4; time++){
            for(char ch: instructions.toCharArray()){
                if(ch == 'L') d = (d + 3) % 4;
                else if(ch == 'R') d = (d + 1) % 4;
                else {
                    x += directions[d][0];
                    y += directions[d][1];
                }
            }

            if(x == 0 && y == 0) return true;
        }

        return false;
    }
}

